src/streams.h

```diff
@@ -15,32 +15,40 @@
 #include <assert.h>
 #include <cstddef>
 #include <cstdio>
+#include <cstring>
 #include <ios>
 #include <limits>
 #include <optional>
+#include <span>
 #include <stdint.h>
 #include <string.h>
 #include <string>
-#include <utility>
 #include <vector>
 
 namespace util {
 inline void Xor(Span<std::byte> write, Span<const std::byte> key, size_t key_offset = 0)
 {
-    if (key.size() == 0) {
-        return;
+    if (write.empty() || key.empty()) return;
+
+    size_t i = 0;
+    // Consume excess key_offset prefix
+    for (key_offset %= key.size(); key_offset > 0 && i < std::min(key.size() - key_offset, write.size()); ++i) {
+        write[i] ^= key[key_offset + i];
+    }
+
+    if (key.size() == 8 && write.size() - i >= 8) { // Xor in 64-bit chunks
+        uint64_t key64;
+        std::memcpy(&key64, key.data(), 8);
+        for (; i <= write.size() - 8; i += 8) {
+            uint64_t write64;
+            std::memcpy(&write64, write.data() + i, 8);
+            write64 ^= key64;
+            std::memcpy(write.data() + i, &write64, 8);
+        }
     }
-    key_offset %= key.size();
-
-    for (size_t i = 0, j = key_offset; i != write.size(); i++) {
-        write[i] ^= key[j++];
 
-        // This potentially acts on very many bytes of data, so it's
-        // important that we calculate `j`, i.e. the `key` index in this
-        // way instead of doing a %, which would effectively be a division
-        // for each byte Xor'd -- much slower than need be.
-        if (j == key.size())
-            j = 0;
+    for (size_t j = 0; i < write.size(); ++i, ++j) {
+        write[i] ^= key[j % key.size()];
```

Comment:
Please avoid % (especially with a dynamic value) in an inner loop. It's essentially a division operation and those are not cheap.

Reply:
Thanks for the hint, I deliberately removed that optimization (please check the commit messages for details), since these are optimized away.
Also, this is just the leftover part, so for key of length 8 (the version used in most places) this will have 7 iterations at most.
Can you see any difference with any of the benchmarks?

Reply:
It's only up to 7 iterations (assuming the key size is 8), sure, youre're right.

But ok, yea, i'm a bit divided about relying on specific non-trivial things being optimized out, makes the output very dependent on specific compiler decisions (which may be fickle in some cases).

Reply:
Often the simplest code gets optimized most, since it's more predictable.
Would you like me to extend the test or benchmark suite or try something else to make sure we're comfortable with the change?

Reply:
No, it's fine. It just gives me goosebumps seeing code like this, but if it doesn't affect the benchmark and no one else cares then it doesn't matter.

Reply:
To get rid of the goosebumps I'm handling the remaining 4 bytes as a single 32 bit xor now, so the final loop (when keys are 8 bytes long, which is mostly the case for us, I think) does 3 iterations at most. So even if it's not optimized away, we should be fine doing 3 divisions by a nice round number like 8.

Reply:
Is that important for max 3 divisions?

Reply:
Usually these optimizations concentrate on the measurable parts based on the profiling results that I'm getting during reindexing or IBD. Obfuscating a single bit (i.e. `XorSmall`) wasn't my focus, it's already very fast, didn't seem like the bottleneck.
Would you like me to concentrate on that scenario instead? Or would it make more sense to serialize a block and use that as the basis for the benchmarks?

> C++ compiler .......................... AppleClang 16.0.0.16000026

Before:
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.99 |      501,740,412.05 |    0.5% |     10.27 | `AutoFileXor`
|                1.24 |      807,597,761.92 |    0.0% |     11.01 | `Xor`
|                1.29 |      776,238,564.27 |    0.0% |     10.59 | `XorSmall`

After:
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                0.73 |    1,364,622,484.81 |    0.9% |      8.78 | `AutoFileXor`
|                0.02 |   40,999,383,920.82 |    0.0% |     11.00 | `Xor`
|                0.54 |    1,862,525,472.57 |    0.0% |     11.00 | `XorSmall`

> C++ compiler .......................... GNU 12.2.0

Before:
|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                1.60 |      624,562,742.35 |    0.2% |            9.20 |            3.57 |  2.579 |           1.03 |    0.1% |     10.61 | `AutoFileXor`
|                0.91 |    1,102,205,071.31 |    0.0% |            9.02 |            3.26 |  2.763 |           1.00 |    0.1% |     11.00 | `Xor`
|                1.23 |      811,876,565.33 |    0.1% |           14.60 |            4.43 |  3.295 |           1.80 |    0.0% |     10.56 | `XorSmall`

After:
|             ns/byte |              byte/s |    err% |        ins/byte |        cyc/byte |    IPC |       bra/byte |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|                0.74 |    1,346,798,809.87 |    0.1% |            0.72 |            0.47 |  1.531 |           0.16 |    0.2% |     11.02 | `AutoFileXor`
|                0.09 |   11,450,472,586.50 |    0.0% |            0.59 |            0.31 |  1.882 |           0.14 |    1.9% |     10.82 | `Xor`
|                5.65 |      177,092,223.60 |    0.1% |           22.00 |           20.31 |  1.083 |           4.80 |    0.0% |     10.99 | `XorSmall`

Reply:
>  It would be good to share the number you go

The `reindex-chainstate` until 600k, 2 runs just finished - comparing master against the 64/32 bit packing (current state) on Linux (with GCC, showing the above inconsistency).

<details>
<summary>Details</summary>

```bash
hyperfine \
--runs 2 \
--show-output \
--export-json /mnt/my_storage/reindex-xor.json \
--parameter-list COMMIT dea7e2faf1bc48f96741ef
84e25e6f47cefd5a92,353915bae14b9704a209bc09b021d3dd2ee11cf2 \
--prepare 'cmake -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_UTIL=OFF -DBUILD_TX=OFF -DBUILD_TESTS=OFF -DENABLE_WALLET=OFF -DI
NSTALL_MAN=OFF && cmake --build build -j$(nproc)' \
'COMMIT={COMMIT} ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=600000 -dbcache=500 -printtoconsole=0 -reindex-chainstate -connect=0'
```

</details>

Before:
* 3.554 hours
* 3.567 hours

After:
* 3.523 hours
* 3.527 hours

<details>
<summary>Details</summary>

Benchmark 1: COMMIT=dea7e2faf1bc48f96741ef84e25e6f47cefd5a92 ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=600000 -dbcache=500 -printtoconsole=0 -reindex-chains
tate -connect=0
  Time (mean ± σ):     12819.367 s ± 35.155 s    [User: 11992.168 s, System: 2509.200 s]
  Range (min … max):   12794.508 s … 12844.225 s    2 runs

Benchmark 2: COMMIT=353915bae14b9704a209bc09b021d3dd2ee11cf2 ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=600000 -dbcache=500 -printtoconsole=0 -reindex-chains
tate -connect=0
  Time (mean ± σ):     12685.350 s ± 19.878 s    [User: 11918.349 s, System: 2523.819 s]
  Range (min … max):   12671.295 s … 12699.406 s    2 runs

</details>

----

Reindexing is a lot more stable than IBD (as seen from multiple measurements), showing a consistent 1% speedup.
Not earth shattering, but at least this way the obfuscation won't cause a speed regression.

----

> Adding a bench for the block couldn't hurt, but I haven't checked how representative it is

Let's find out!

Reply:
I've change the usages to avoid std::vector keys, this way GCC and Clang both agree that the new results are faster (even though clang manages to compile to 32 byte SIMD, while GCC only to 16 bytes per iteration, see https://github.com/bitcoin/bitcoin/pull/31144#discussion_r1817939675)